### PR TITLE
`ilab model evaluate` command and eval library usage

### PR DESCRIPTION
**Pull request by cdoern**
1) added support for `ilab model evaluate` which allows users to run MMLU Bench MT Bench, MMLU Branch, and MT Branch Benchmarks
2) Add an `_evaluate` class to `config.yaml` so that users can get sane evaluation defaults that they can see and modify. These funnel directly into the evaluation flags as training now does. 
a sample evaluation class looks like:
```
evaluate:
  branch: null
  mmlu:
    batch_size: 5
    few_shots: 2
  mmlu_branch:
    sdg_path: generated
  model_name: null
  mt:
    judge_model: prometheus-eval/prometheus-8x7b-v2.0
    max_workers: 40
    output_dir: eval_data
  mt_branch:
    taxonomy_path: taxonomy
```

### CODE REVIEW

**Comment by cdoern** [comment]
this is a placeholder for now, lmk when the library is somewhere I can access and import

**Comment by nathan-weinberg** [comment]
@cdoern you can install directory from test.pypy.org for testing if you wish: https://test.pypi.org/project/instructlab-eval

once we have a `0.0.1` release we'll publish that to production PyPI

File: src/instructlab/model/evaluate.py
Comment (leseb) [inline src/instructlab/model/evaluate.py:29]: is it only for vLLM? How about llama-ccp?

File: src/instructlab/model/evaluate.py
Comment (leseb) [inline src/instructlab/model/evaluate.py:55]:
```suggestion
    help="List of tasks for mmlu to test the model with. Needed for running mmlu.",
```

File: src/instructlab/model/evaluate.py
Comment (leseb) [inline src/instructlab/model/evaluate.py:70]:
```suggestion
@utils.display_params
def evaluate(
```

File: src/instructlab/model/evaluate.py
Comment (leseb) [inline src/instructlab/model/evaluate.py:109]: Can we have this in a dedicated function so that it can be unit tested accordingly, that's a lot of conditions.

**Review by leseb — CHANGES_REQUESTED** [review]

File: src/instructlab/model/evaluate.py
Comment (cdoern) [reply src/instructlab/model/evaluate.py:70]: is this new? What does this do?

Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:70]: Curious as well

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:109]: Yeah think we could do something like `selectEvaluator()` where we pass the benchmark and return the proper class if all the needed args are present, @cdoern I can look into doing this if you want

Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py]: @leseb did something like this in my latest commit

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py]: MMLU is an acronym, so it should be capitalized

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py]: Arg has been removed in subsequent configs, but also I believe vLLM is required - @alimaredia can you confirm?

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:75]: elif

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:95]: Name of model to be evaluated?

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:108]: Should this have a default?

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:94]: No longer a path

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:105]: No longer a path

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:106]: Name of the model....

**Review by danmcp — CHANGES_REQUESTED** [review]

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:75]: Any particular reason? I don't see a functional difference

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:95]: Where will it look for the model once it has the name?

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:94]: Same question as above

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:106]: Same question as above...

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:105]: Same question as above

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:95]: For mtbench, it looks for the model being served on server_url.  The mmlu work hasn't switched to looking up the hosted model yet.  My comment was just meant to point out the description didn't match the option name.

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:75]: I would suggest it makes the code more clear (that it will never hit the second if if the first one succeeds) and it avoid the extra unneeded string compare in those cases.

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:108]: I don't know.  Consistency with other cli dirs would be good.  The other consideration is whether we want to advertise the output to users or not.  Currently there isn't a reason to point users to it so I would suggest a . directory.

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:95]: @alinaryan ^^ FYI

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:108]: Spoke with @cdoern, adding a default output dir called `eval_results` :+1:

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py]: > `elif` would still require the string compare no?

the elif wouldn't compare if the first if was a match

File: src/instructlab/configuration.py
Comment (danmcp) [inline src/instructlab/configuration.py:32]: results may not be the best term here.  The folder will contain a lot of the intermediate states.  The actual results should probably be returned through the command.

Comment (nathan-weinberg) [reply src/instructlab/configuration.py:32]: Going to change to `eval_data` - @RobotSail this aligns with what's in your proposal, though the pathing will need to be changed later

Comment (RobotSail) [reply src/instructlab/configuration.py]: @nathan-weinberg Yes this is perfect! @danmcp We will move this into a more reasonable location later.

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [inline src/instructlab/model/evaluate.py:154]: Indentation is off

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [inline src/instructlab/model/evaluate.py:160]: why is this a single-item list?

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:160]: It can be multiple items.  But we are only going to use it for the whole group.

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:154]: Apologies, I probably missed a format before pushing.  Giving we are sharing the pull, I'll wait till my next commit to fix.

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:160]: Gotcha!

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:154]: No worries!

**Comment by cdoern** [comment]
I think CI should just pass on this one :crossed_fingers:

**Review by nathan-weinberg — DISMISSED** [review]

**Review by nathan-weinberg — APPROVED** [review]

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:289]: Noting this is a WIP here: https://github.com/instructlab/instructlab/compare/main...danmcp:instructlab:evalandvllm?expand=1

Comment (danmcp) [inline src/instructlab/model/evaluate.py:289]: Noting this is a WIP here: https://github.com/instructlab/instructlab/compare/main...danmcp:instructlab:evalandvllm?expand=1

File: src/instructlab/model/evaluate.py
Comment (danmcp) [inline src/instructlab/model/evaluate.py:395]: This server launch could be shared between the two models.  Addressed here: https://github.com/instructlab/instructlab/compare/main...danmcp:instructlab:evalandvllm?expand=1

**Review by danmcp — APPROVED** [review]
I am approving with a few notes:
- The error handling still needs a decent bit of work both here and the eval library.  -> Replace stack traces with nice messages.
- This version is very chatty in the cli.  Will largely be addressed with https://github.com/instructlab/instructlab/compare/main...danmcp:instructlab:evalandvllm?expand=1.  Those commit(s) also address the vllm serving hack still in place in this PR.
- More test cases are needed
- Config defaulting probably still needs some work with the models selected

File: src/instructlab/model/evaluate.py
Comment (alinaryan) [inline src/instructlab/model/evaluate.py:71]: Should `test_model` be `model`? Applies here and other places too

File: src/instructlab/model/evaluate.py
Comment (alinaryan) [inline src/instructlab/model/evaluate.py:98]: Are you planning to reference the [default 57 tasks](https://github.com/instructlab/eval/pull/24/files#diff-88873e8f609db17f8729f4725117acdfa0db7930066436bd6b0d5c7301511101R14) in this PR?

File: src/instructlab/model/evaluate.py
Comment (alinaryan) [inline src/instructlab/model/evaluate.py:100]: Is this needed as some reinforcement? If so, should you hard code the full list of 57 tasks here?

File: src/instructlab/model/evaluate.py
Comment (alinaryan) [inline src/instructlab/model/evaluate.py:219]: don't forget to add the default `eval_results`

**Review by alinaryan — COMMENTED** [review]
Thanks for the strong work on this!! Just have a few comments/q's:

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:100]: INSTRUCTLAB_EVAL_MMLU_MIN_TASKS is in place for testing purposes to allow for a smaller list of tasks to validate the flow without running a full test.

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:219]: The default is set to eval_data in configuration.py

File: src/instructlab/model/evaluate.py
Comment (danmcp) [reply src/instructlab/model/evaluate.py:71]: In this case test_model is simply the name referenced by mt_bench to call the model being served.  It could be anything so we just want to pick something recognizable.  And calling the model you are serving "model" seems a little under descriptive.

Comment (alinaryan) [reply src/instructlab/model/evaluate.py:71]: Sounds good to me! Wanted to ensure it wasn't missed with the other renames to 'model'

File: src/instructlab/model/evaluate.py
Comment (alinaryan) [reply src/instructlab/model/evaluate.py:219]: Ah. Thought it had to be added in this click arg

**Review by alinaryan — APPROVED** [review]

File: src/instructlab/model/evaluate.py
Comment (alinaryan) [reply src/instructlab/model/evaluate.py:98]: Spoke with @cdoern - this will be addressed in a follow-up PR. See issue https://github.com/instructlab/instructlab/issues/1534

**Comment by cdoern** [comment]
made follow up issues!

File: src/instructlab/configuration.py
Comment (cdoern) [reply src/instructlab/configuration.py:429]: since `_evaluate` has "top level" options that arent in sub classes, the `|` operator can't pull out primitive values (str, float, etc). So, what I do here is create a map with the entire old eval class, AND I add the flattened sub classes. so it looks something like 

` eval
    mt:
    mmlu
    mmlubranch
    mtbranch
    ALL_OPTIONS_FROM_THOSE_CLASSES_DUPLICATED_IN_TOP_LEVEL
`
I then delete mt, mmlu, and the branch benches so I am left with just the "unfolded" options in a "top level" eval key that can be directly mapped to the flags.

This is super confusing. I chatted with @nathan-weinberg about it, I can whiteboard it for folks on monday too

File: src/instructlab/configuration.py
Comment (cdoern) [reply src/instructlab/configuration.py:416]: yep, for training yeah. I need to split out each cmd that we want to auto map

File: src/instructlab/configuration.py
Comment (cdoern) [reply src/instructlab/configuration.py:182]: Making a follow up issue for this since I don't want to lose the two approvals we currently have with a small push

File: src/instructlab/configuration.py
Comment (alinaryan) [reply src/instructlab/configuration.py:429]: That would be helpful, thank you :pray:

**Comment by cdoern** [comment]
Given that tomorrow morning (7/1) is a deadline. This PR needs to be merged before then to add some form of evaluation support.

That being said, there is a stale change request on this PR, some pending reviews that haven't come in yet, etc.

We will be dismissing those in favor of deferring to follow up issues but if any of the reviewers have immediate follow up concerns please feel free to reach out!!!!

**Comment by russellb** [comment]
I filed #1540 as a follow-up to get this tested in `e2e` CI

File: src/instructlab/model/evaluate.py
Comment (nathan-weinberg) [reply src/instructlab/model/evaluate.py:100]: full list is also now a default of the MMLU evaluator class as well, so we can initialize it w/o that arg to run full-scale MMLU: https://github.com/instructlab/eval/pull/24

**Pull request merged**
